row_generation/row_generator.py

```diff
@@ -21,18 +25,24 @@ def __init__(self, stage: int):
         self._row = self.rounds()
 
     def reset(self):
+        """ Reset the row generator. """
+
         self.logger.info("Reset")
         self._has_bob = False
         self._has_single = False
         self._index = 0
         self._row = self.rounds()
 
     def reset_calls(self):
+        """ Clear the pending call flags. """
+
         self.logger.info("Reset calls")
         self._has_bob = False
         self._has_single = False
 
     def next_row(self, is_handstroke: bool) -> List[Bell]:
+        """ Generate the next row, and mutate state accordingly. """
```

Comment:
As an aside, I keep wondering if it would be nice for this to be stateless so multiple consumers could call this

* Essentially make this into `_gen_row()` so
  1) Replace self._row with `previous_row`
  1) Replace `self._index` with a `row_count` argument, and rely on the consumer to be counting from the start (i.e. 0 at rounds)
* Truly stateless calling would be harder
  1) It would need an argument of current calls (i.e. has a bob been called)
  1) Return an indication of whether to clear the calls
  1) Some implementations keep track of other state, e.g. `PlaceNotationGenerator` to keep track of calls which affect multiple rows

Reply:
I think that making it stateless is an anti-pattern (at least coming from Rust), since any RowGenerator is essentiall is an iterator that produces rows.

In Rust, an Iterator is any struct that implements a method `next`, which takes a mutable reference to the iterator and returns either `Some(next_element)`, or `None` to say that the iterator is finished.  Usually these are impossible to pass by reference (and get cloned if you need to use them multiple times), so the mutability/statefullness is neatly contained.  However, this is Python so this isn't enforcable, but I still think that it's a good way to think about iterators.

See the [rust iterator documentation](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next).

Reply:
I think making it stateless would 
1) Make some things much harder to implement
2) Just push the state elsewhere (i.e. where the calls are)

Possibly we could expose `_row` as `current_row` if there would be another consumer.